cache manifests on pull

cpuguy83 opened the pull request:
The tag service does a `HEAD` request to get the manifest digest, where
we can then do a `GET` against the digest.

The `GET` by tag is not cacheable, but the `GET` against the digest is.
This allows proxies to work way better.

This is also how containerd works.

It also caches the manifest in containerd's content store so if `HEAD` returns a manifest we already have we don't need to hit the registry again.
Caches are cleaned up when all images using that manifest are removed.

It also saves a round trip on pull by digest when the manifest is already cached.

### Before this change, mitmdump:

```
127.0.0.1:47070: clientconnect
127.0.0.1:47070: GET https://registry-1.docker.io/v2/
              << 401 Unauthorized 87b
127.0.0.1:47070: clientdisconnect
127.0.0.1:47072: clientconnect
127.0.0.1:47072: GET https://auth.docker.io/token?scope=repository%3Alibrary%2Fbusybox%3Apull&service=registry.docker.io
              << 200 OK 4.25k
127.0.0.1:47074: clientconnect
127.0.0.1:47072: clientdisconnect
127.0.0.1:47074: GET https://registry-1.docker.io/v2/library/busybox/manifests/latest
              << 200 OK 2.03k
127.0.0.1:47074: clientdisconnect
127.0.0.1:47076: clientconnect
127.0.0.1:47076: GET https://registry-1.docker.io/v2/library/busybox/manifests/sha256:c9249fdf56138f0d929e2080ae98ee9cb2946f71498fc1484288e6a935b5e5bc
              << 200 OK 527b
127.0.0.1:47076: clientdisconnect
```


### With this change, mitmdump:

```
127.0.0.1:46284: clientconnect
127.0.0.1:46284: GET https://registry-1.docker.io/v2/
              << 401 Unauthorized 87b
127.0.0.1:46284: clientdisconnect
127.0.0.1:46286: clientconnect
127.0.0.1:46286: GET https://auth.docker.io/token?scope=repository%3Alibrary%2Fbusybox%3Apull&service=registry.docker.io
              << 200 OK 4.25k
127.0.0.1:46290: clientconnect
127.0.0.1:46286: clientdisconnect
127.0.0.1:46290: HEAD https://registry-1.docker.io/v2/library/busybox/manifests/latest
              << 200 OK 0b
127.0.0.1:46290: clientdisconnect
127.0.0.1:46292: clientconnect
127.0.0.1:46292: GET https://registry-1.docker.io/v2/library/busybox/manifests/sha256:a9286defaba7b3a519d585ba0e37d0b2cbee74ebfe590960b0b1d6a5e97d1e1d
              << 200 OK 2.03k
127.0.0.1:46292: clientdisconnect
127.0.0.1:46294: clientconnect
127.0.0.1:46294: GET https://registry-1.docker.io/v2/library/busybox/manifests/sha256:c9249fdf56138f0d929e2080ae98ee9cb2946f71498fc1484288e6a935b5e5bc
              << 200 OK 527b
127.0.0.1:46294: clientdisconnect
```

cpuguy83 committed:
pull: use tag service for pulling tagged reference

The tag service does a `HEAD` request to get the manifest digest, where
we can then do a `GET` against the digest.

The `GET` by tag is not cacheable, but the `GET` against the digest is.
This allows proxies to work way better.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

tianon commented on distribution/pull_v2.go:
```diff
@@ -346,10 +346,16 @@ func (p *v2Puller) pullV2Tag(ctx context.Context, ref reference.Named, platform
 			return false, allowV1Fallback(err)
 		}
+
+		manifest, err = manSvc.Get(ctx, desc.Digest)
```
Aren't there situations where Docker ends up storing the digest of the fetched image?  Would that make sense to check here in case it's not a manifest list?  (Does this function already do so? 😅)

estesp replied:
The switch on `manifest.(type)` happens just a little further down, if I'm understanding your question; if it's a manifest list at this point it will call the pullManifestList function

cpuguy83 replied:
Yeah, this is handled later on.

tonistiigi commented:
The reason containerd library does this is that the manifest can be available locally and `GET` can be skipped (eg. buildkit implements it that way). Moby doesn't save the manifest atm (outside buildkit) and this just results in an extra request that adds no additional information.

cpuguy83 commented:
Absolutely right, we aren't storing the manifest locally yet, so we need to do a round trip each time... but...

I can't (effectively) use a caching proxy for manifest by tag.
With this change Docker always  gets the manifest by digest, which I can cache.

With rate limiting going into effect it this also makes a huge difference since HEAD does not count against the limit.

tonistiigi commented:
@cpuguy83 Atm you are suggesting making everyone's pulls slower for the use case you will have if you implement your proxy in a certain way. I'm sure there will be more people who will find this useful but it is still a niche case. Things would be different of course if we could improve the experience for other users as well, and extra HEAD would not just add extra time to their pulls but provide better value through cache.

cpuguy83 commented:
@tonistiigi 

That's one way to put it...

I would put it as, sacrificing an extra round trip for the possibility of wrapping docker in a caching proxy.
It's not about "implementing your proxy in a certain way".
`GET /v2/repo/manifests/latest` -> not cachable
`GET /v2/repo/manifests/<diges>` -> cachable
It only creates any meaningful slowdown (in terms of % of the entire pull) when you already have the image locally with the benefit of being able to fetch the manifest from a (dumb) local proxy.

cpuguy83 committed:
Store image manifests in containerd content store

This allows us to cache manifests and avoid extra round trips to the
registry for content we already know about.

dockerd currently does not support containerd on Windows, so this does
not store manifests on Windows, yet.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
I updated this to use containerd to store fetched manifests.

I'm not sure how to clean these manifests out.
Maybe it's not important given how small they are (until it fills up someone's prod machine)

cpuguy83 commented:
This stores both the manifest list and the actual fetched manifest.

tonistiigi commented:
> I'm not sure how to clean these manifests out.

There are two ways to clean containerd objects, with leases or with root labels. Leases is the preferred option. We could create a lease per image, add manifest(s) to that lease and remove the lease together with the image. This acts as a ref count so when last least pointing to manifest goes away then manifest is released. The process is similar for the root labels. Even when manifest is already cached it would be added to the new lease - this is the case when one multi-arch manifest can be part of multiple local images.

Note that buildkit integration uses its own contentstore atm (but exactly the same interfaces). This was done to not have any potential backward compatibility issues on actual migration to containerd storage. If we are comfortable to use the containerd daemon for this we can update buildkit code to also switch to this contentstore (and migrate the existing data). Or we can use the local moby-only instance as well like buildkit does atm. Bottom line is that they should start to use the same store but that's not blocking this PR.

cpuguy83 commented:
Updated this to handle deletes using leases.

During an active pull we add a temporary lease to prevent content from being accidentally deleted by GC.
After the pull is complete, we add a persistent lease keyed on the image ID.

The lease is deleted when the image is deleted, after which the manifests are collected by containerd's GC.

Nothing is persisted on Windows since we don't support containerd on Windows yet.

amouat commented:
This is suggesting changes to the Distribution Specification, right? I know you're trying to mitigate a problem that's happening right now, but it might be an idea to start a conversation over there https://github.com/opencontainers/distribution-spec

cpuguy83 commented:
Right. We discussed this on the last spec call that HEAD is missing detail.

amouat commented:
Thanks for the clarification. I should have checked first and I would have found there is even an open issue for this: https://github.com/opencontainers/distribution-spec/issues/207

cpuguy83 commented:
And added a 3rd commit to create a local content/metadata store to handle this on Windows, where we don't use containerd yet.

cpuguy83 commented on distribution/pull_v2.go:
```diff
@@ -43,6 +47,10 @@ var (
 )
 
+const (
+	labelMediaType = "com.docker.distribution.mediaType"
```
@tonistiigi WDYT about this label key?

tonistiigi replied:
I still don't quite understand why you need it. You already call the detection function and that seems fine to me. The labeling part isn't reliable anyway for all of the cases, I'd just completely remove it.

cpuguy83 commented:
This should be ready.
I'd like some more tests around the lease handling but this is fairly difficult to mock (not because of leases, difficulty is our mess).
I'll look into this more.

Added unit tests for the local manifest caching.
And the last test run was all green on CI (last push was some logging changes + unit tests).

tonistiigi commented on daemon/images/image_pull.go:
```diff
@@ -124,3 +155,130 @@ func (i *ImageService) GetRepository(ctx context.Context, ref reference.Named, a
+	return id, s.updateLease(id)
+}
+
+func (s *imageStoreWithLease) Get(id image.ID) (*image.Image, error) {
```
iiuc this `Get/Delete` is only used on pull and not in regular image operations. Would be good to leave comments about what conditions they actually get called.

tonistiigi commented on daemon/images/image_delete.go:
```diff
@@ -73,7 +77,25 @@ func (i *ImageService) ImageDelete(imageRef string, force, prune bool) ([]types.
+
 	repoRefs := i.referenceStore.References(imgID.Digest())
+	defer func() {
+		// We only want to delete the lease if this is the last reference to this image
```
Wouldn't it make sense to connect this with the imageStore level and delete lease on `imageStore.Delete()`.

cpuguy83 replied:
Good call, much simpler.

tonistiigi commented on daemon/images/image_pull.go:
```diff
@@ -124,3 +155,130 @@ func (i *ImageService) GetRepository(ctx context.Context, ref reference.Named, a
+	return nil
+}
+
+func tempLease(ctx context.Context, mgr leases.Manager) (context.Context, func(context.Context) error, error) {
```
btw, buildkit sets a special label for leases like this https://github.com/moby/buildkit/blob/master/util/leaseutil/manager.go#L34 just in case. Not a requirement but might want to do the same.

tonistiigi commented on daemon/images/image_pull.go:
```diff
@@ -124,3 +155,130 @@ func (i *ImageService) GetRepository(ctx context.Context, ref reference.Named, a
+}
+
+func imageKey(id image.ID) string {
+	return "image-" + id.String()
```
`moby-image-` ?

This should already be namespaced so not super important but in a follow up we should figure out how to make sure every dockerd is on its own isolated namespace. Eg. if 2 dockerds share same containerd they would now start to delete each others objects.

cpuguy83 replied:
We already have daemon flags b/c of this for integration tests.

tonistiigi replied:
Yes, but atm it is kind of optional and unlikely to cause damage I believe. Anyway, other than the prefix this is not in the scope of this PR anyway.

cpuguy83 commented:
Updated this with suggested changes + re-organized some things.

tonistiigi commented on daemon/images/image_delete.go:
```diff
@@ -60,7 +60,7 @@ const (
 // conflict will not be reported.
 //
-func (i *ImageService) ImageDelete(imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error) {
+func (i *ImageService) ImageDelete(imageRef string, force, prune bool) (_ []types.ImageDeleteResponseItem, retErr error) {
```
nit: looks like this file doesn't have meaningful changes after the updates

tonistiigi commented on daemon/images/store.go:
```diff
@@ -0,0 +1,145 @@
+
+func (c *contentStoreForPull) Writer(ctx context.Context, opts ...content.WriterOpt) (content.Writer, error) {
+	w, err := c.ContentStore.Writer(ctx, opts...)
+	if err != nil {
```
https://github.com/moby/moby/pull/41607#discussion_r516421816

cpuguy83 replied:
Fixed.

thaJeztah commented:
/cc @simskij

thaJeztah commented:
@AkihiroSuda @estesp ptal 🤗

cpuguy83 commented:
Updated this to fix some typos, and found a bug where it wasn't closing the `ReaderAt` in the case where it needs to detect the media type (pull by digest) of local content, so fixed that.

thaJeztah commented:
Odd failure on ppc64le https://ci-next.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-41607/runs/39/nodes/51/log/?start=0

```
`=== FAIL: daemon TestRootMountCleanup/regular_dir_no_mountpoint (0.02s)
    --- FAIL: TestRootMountCleanup/regular_dir_no_mountpoint (0.02s)
        daemon_linux_test.go:216: assertion failed: expression is false: mounted == expect: expected false, actual true
        daemon_linux_test.go:219: assertion failed: expression is false: os.IsNotExist(err)
```

s390x failures are the flaky ones https://github.com/moby/moby/issues/41561

thaJeztah commented:
I guess the failure on ppc64 could be a race; it's failing after `daemon.cleanupMounts()` is called https://github.com/moby/moby/blob/af0415257efa288d9444de46ba2b45868c7c1a54/daemon/daemon_linux_test.go#L215-L216

which lazily unmounts; https://github.com/moby/moby/blob/af0415257efa288d9444de46ba2b45868c7c1a54/vendor/github.com/moby/sys/mount/mount.go#L20-L21


```
// Unmount lazily unmounts a filesystem on supported platforms, otherwise does
// a normal unmount.
```

So I guess that test could check if it's mounted when it's still in the process of being unmounted? Not sure

gabrielcalderon commented:
Do these changes also affect the Docker Registry? Specially now with the new pull rate limits?

cpuguy83 commented:
Yes.
A HEAD request does not count towards the rate limit.
If the manifest is already cached then this saves a GET to the registry, which does affect rate limit.

gabrielcalderon commented:
Great to hear! Hopefully these changes are also integrated into the Registry itself. It seems when it runs in Mirror mode, it checks upstreams for new manifests.

KevinCathcart commented:
@gabrielcalderon I've seen that claim a bunch, but I'm a bit confused by it when checking the registry code over in docker/distribution. The code for the pull-though cache mode, caches manifests when fetched by digest. When you try to fetch by tag, it uses TagService to resolve the tag to a manifest digest, and then pulls via that digest in order to cache it. 

Indeed, this pull request is quite literally making dockerd code more like docker/distribution's pull through cache mode, using TagService to avoid the rate limit penalty, plus adding a local manifest cache keyed by the digest.

It does look like the registry in pull though cache mode will only cache manifests for 7 days. I'm assuming this is the mechanism by which it prevents the cache from growing over time until the disk is full. This is not well documented, and is likely a much shorter time frame than some people are expecting.

I just tested this locally, pulling `alpine:latest`'s manifest from a local registry in pull through cache mode a whole bunch of times, and it did not change my rate limit after the first manifest GET. 

I suspect most of the people saying that the registry's pull though cache does not protect against the rate limit are actually expecting that it caches manifests for more than 7 days (I know I was expecting a much longer cache period), or were hitting a bunch of images where the tag had a new version pushed.

thaJeztah commented:
@AkihiroSuda @tiborvass are you able to help review?

gabrielcalderon commented:
@KevinCathcart That's great to hear. The logging messaged in the registry are not very clear about wether it's using Local or pulling from remote to update the cache.

Totally agree on the lack of full documentation of that behavior. The official documentation just says that it checks Docker Hub for new content.
